From ff0219c19b4172235e41e302a58dd47cb1d3a28a Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 11 Apr 2022 12:56:22 +0300
Subject: [PATCH] [SW-66329] module: pci: move source file to lib common

Pci module is no longer used by preboot application only, in the next
patch it will also be used by the pif arc application. so it's being
moved to the common lib folder

Change-Id: Icd44f6a7b19e7e631ea4b96652185e0d8cc11fc7
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index b224300..cfdcb7a 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -61,6 +61,7 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/string.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/lib_common.o \
+	$(PREBOOT_ROOT_DIR)/lib/common/module_pci.o \
 	$(PREBOOT_ROOT_DIR)/app/preboot/preboot.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/boot_info.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/fit_regions.o \
@@ -68,7 +69,6 @@
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_uart.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/fit_rsa.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_i2c_slave.o \
-	$(PREBOOT_ROOT_DIR)/lib/preboot/module_pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/crc8.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/crc32.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/i2cslave_indirect.o \
diff --git a/lib/common/module_pci.c b/lib/common/module_pci.c
new file mode 100644
index 0000000..fe8124b
--- /dev/null
+++ b/lib/common/module_pci.c
@@ -0,0 +1,264 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2016-2020 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+
+#include <bitops.h>
+#include <preboot.h>
+#include <device.h>
+
+#include "cmd_module.h"
+#include "uart.h"
+#include "pci.h"
+#include "pll.h"
+#include <hl.h>
+
+struct pci_status {
+	u32 flr_reg_val;
+	u32 hot_rst_reg_val;
+	u32 smlh_link_reg_val;
+	u32 rdlh_link_reg_val;
+	bool link_is_up;
+	bool mem_enabled;
+	bool mem_enabled_prev;
+	bool pci_pll_initialized;
+	bool setup_done;
+};
+
+static struct pci_status pci_sts = {
+	.flr_reg_val = ~1,
+	.hot_rst_reg_val = ~1,
+	.smlh_link_reg_val = 1,
+	.rdlh_link_reg_val = 1,
+	.link_is_up = false,
+	.mem_enabled_prev = false,
+	.pci_pll_initialized = false,
+	.setup_done = false,
+};
+
+__weak void pci_dev_specific_setup(void) {}
+__weak void pci_dev_specific_teardown(void) {}
+__weak void pci_dev_specific_disable(void) {}
+__weak void pci_dev_specific_enable(void) {}
+
+#define PCIE_AUX_FLR_CTRL_HW_CTRL_MASK	0x1
+#define PCIE_AUX_FLR_CTRL_INT_MASK_MASK	0x2
+
+__weak void pci_dev_flr_enable(bool enable)
+{
+	u32 val = enable ? 0 : PCIE_AUX_FLR_CTRL_HW_CTRL_MASK |
+				PCIE_AUX_FLR_CTRL_INT_MASK_MASK;
+
+	regmap_writel(val, pci.aux_flr_ctrl_reg);
+}
+
+__weak void pci_dev_specific_hot_rst(void)
+{
+	regmap_writel(0x10, pci.core_hot_rst_reg);
+	regmap_writel(1, pci.aux_ltssm_en_reg);
+}
+
+__weak void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, pci.aux_pf_flr_active_reg);
+	regmap_writel(1, pci.aux_flr_int_reg);
+	regmap_writel(1, pci.aux_pf_flr_done_reg);
+}
+
+__weak void hl_pci_iatu_init(void) {}
+__weak bool is_pcie_initialized(void) { return true; }
+__weak void hl_dev_pci_pll_wa(void) {}
+
+static void module_pci_do_setup(void)
+{
+	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
+		pci_dev_specific_setup();
+		pci_dev_flr_enable(true);
+	}
+
+	pci_sts.setup_done = true;
+}
+
+void module_pci_setup(void)
+{
+	if (!is_pcie_initialized())
+		return;
+
+	module_pci_do_setup();
+}
+
+#define DBG_TIMEOUT_US		1000000
+#define DBG_LINE_SZ		80
+
+static unsigned int dbg_cnt;
+static u64 dbg_ts;
+
+static void pci_init_wait_dbg(void)
+{
+	if (timeout_expired(dbg_ts, DBG_TIMEOUT_US)) {
+		preboot_puts_raw_dbg("*");
+		dbg_ts = timestamp_get_us();
+		dbg_cnt++;
+	}
+
+	if (dbg_cnt == DBG_LINE_SZ) {
+		preboot_puts_dbg("\n");
+		dbg_cnt = 0;
+	}
+}
+
+__weak u32 pci_dev_specific_hot_rst_check(void)
+{
+	return regmap_readl(pci.core_hot_rst_reg) & BIT(0);
+}
+
+__weak enum cmd_check_res dev_specific_module_pci_check(void)
+{
+	/*
+	 * don't read pci status regs until PRSTn is deasserted and pci blocks
+	 * are functional
+	 */
+	if (!is_pcie_initialized()) {
+		pci_init_wait_dbg();
+		return NOT_FOUND;
+	}
+
+	if (!pci_sts.setup_done)
+		module_pci_do_setup();
+
+	pci_sts.flr_reg_val = regmap_readl(pci.aux_flr_int_reg);
+	pci_sts.hot_rst_reg_val = pci_dev_specific_hot_rst_check();
+	pci_sts.rdlh_link_reg_val = regmap_readl(pci.aux_rdlh_link_up_reg);
+	pci_sts.smlh_link_reg_val = regmap_readl(pci.aux_smlh_link_up_reg);
+
+	if (!pci_sts.pci_pll_initialized)
+		return FOUND_SLOW_RETURN;
+
+	if (pci_sts.flr_reg_val == 1 || pci_sts.hot_rst_reg_val == 1)
+		return FOUND_FAST_RETURN;
+
+	if ((pci_sts.rdlh_link_reg_val && pci_sts.smlh_link_reg_val) !=
+	    pci_sts.link_is_up)
+		return FOUND_FAST_RETURN;
+
+	if (pci_sts.link_is_up)
+		pci_sts.mem_enabled = hl_pci_is_mem_enabled();
+
+	if (pci_sts.mem_enabled != pci_sts.mem_enabled_prev)
+		return FOUND_FAST_RETURN;
+
+	return NOT_FOUND;
+}
+
+enum cmd_check_res module_pci_check(void)
+{
+	if (hl_is_debug_mode(HL_DEBUG_HW_PCIE))
+		return NOT_FOUND;
+
+	return dev_specific_module_pci_check();
+}
+
+void module_pci_disable(void)
+{
+	if (!is_pcie_initialized())
+		return;
+
+	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
+		pci_dev_specific_disable();
+		pci_dev_flr_enable(false);
+	}
+
+	pci_sts.setup_done = false;
+}
+
+__weak void pci_dev_specific_run(void)
+{
+	unsigned int rst_src = hl_rst_src_get();
+
+	if (!pci_sts.pci_pll_initialized) {
+		if ((rst_src == HL_COLD_RST) || (rst_src == HL_PRSTN_RST))
+			hl_dev_pci_pll_wa();
+
+		pci_sts.pci_pll_initialized = true;
+	}
+}
+
+void module_pci_run(void)
+{
+	pci_dev_specific_run();
+}
+
+__weak void pci_dev_specific_post_link_up(void) {}
+
+void module_pci_run_fast(void)
+{
+	unsigned int rst_src;
+
+	if (pci_sts.flr_reg_val == 1) {
+		pci_dev_specific_flr();
+		preboot_puts("FLR received\n");
+	}
+
+	if (pci_sts.hot_rst_reg_val == 1) {
+		pci_dev_specific_hot_rst();
+		preboot_puts("HOT RST received\n");
+	}
+
+	if ((pci_sts.rdlh_link_reg_val && pci_sts.smlh_link_reg_val) !=
+	    pci_sts.link_is_up) {
+		pci_sts.link_is_up = !pci_sts.link_is_up;
+		preboot_puts_raw("PCIe link is ");
+		if (pci_sts.link_is_up) {
+			preboot_puts("up\n");
+			pci_dev_specific_post_link_up();
+		} else {
+			preboot_puts("down\n");
+			pci_sts.mem_enabled = false;
+		}
+	}
+
+	if (pci_sts.mem_enabled_prev != pci_sts.mem_enabled) {
+		pci_sts.mem_enabled_prev = pci_sts.mem_enabled;
+		rst_src = hl_rst_src_get();
+		if (pci_sts.mem_enabled && ((rst_src == HL_COLD_RST) ||
+					    (rst_src == HL_PRSTN_RST)))
+			hl_pci_iatu_init();
+
+		preboot_puts_raw("PCIe mem_en changed to: ");
+		preboot_puts_raw(pci_sts.mem_enabled ? "true" : "false");
+		preboot_puts("\n");
+	}
+}
+
+void module_pci_enable(void)
+{
+	if (!is_pcie_initialized())
+		return;
+
+	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
+		pci_dev_specific_enable();
+		pci_dev_flr_enable(true);
+	}
+
+	pci_sts.setup_done = true;
+}
+
+void module_pci_teardown(void)
+{
+	if (!is_pcie_initialized())
+		return;
+
+	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
+		pci_dev_specific_teardown();
+		pci_dev_flr_enable(false);
+	}
+
+	pci_sts.setup_done = false;
+}
+
+bool hl_pci_link_is_up(void)
+{
+	return pci_sts.link_is_up;
+}
diff --git a/lib/preboot/module_pci.c b/lib/preboot/module_pci.c
deleted file mode 100644
index fe8124b..0000000
--- a/lib/preboot/module_pci.c
+++ /dev/null
@@ -1,264 +0,0 @@
-// SPDX-License-Identifier: GPL-2.0+
-/*
- * Copyright 2016-2020 HabanaLabs, Ltd.
- * All Rights Reserved.
- */
-
-
-#include <bitops.h>
-#include <preboot.h>
-#include <device.h>
-
-#include "cmd_module.h"
-#include "uart.h"
-#include "pci.h"
-#include "pll.h"
-#include <hl.h>
-
-struct pci_status {
-	u32 flr_reg_val;
-	u32 hot_rst_reg_val;
-	u32 smlh_link_reg_val;
-	u32 rdlh_link_reg_val;
-	bool link_is_up;
-	bool mem_enabled;
-	bool mem_enabled_prev;
-	bool pci_pll_initialized;
-	bool setup_done;
-};
-
-static struct pci_status pci_sts = {
-	.flr_reg_val = ~1,
-	.hot_rst_reg_val = ~1,
-	.smlh_link_reg_val = 1,
-	.rdlh_link_reg_val = 1,
-	.link_is_up = false,
-	.mem_enabled_prev = false,
-	.pci_pll_initialized = false,
-	.setup_done = false,
-};
-
-__weak void pci_dev_specific_setup(void) {}
-__weak void pci_dev_specific_teardown(void) {}
-__weak void pci_dev_specific_disable(void) {}
-__weak void pci_dev_specific_enable(void) {}
-
-#define PCIE_AUX_FLR_CTRL_HW_CTRL_MASK	0x1
-#define PCIE_AUX_FLR_CTRL_INT_MASK_MASK	0x2
-
-__weak void pci_dev_flr_enable(bool enable)
-{
-	u32 val = enable ? 0 : PCIE_AUX_FLR_CTRL_HW_CTRL_MASK |
-				PCIE_AUX_FLR_CTRL_INT_MASK_MASK;
-
-	regmap_writel(val, pci.aux_flr_ctrl_reg);
-}
-
-__weak void pci_dev_specific_hot_rst(void)
-{
-	regmap_writel(0x10, pci.core_hot_rst_reg);
-	regmap_writel(1, pci.aux_ltssm_en_reg);
-}
-
-__weak void pci_dev_specific_flr(void)
-{
-	regmap_writel(1, pci.aux_pf_flr_active_reg);
-	regmap_writel(1, pci.aux_flr_int_reg);
-	regmap_writel(1, pci.aux_pf_flr_done_reg);
-}
-
-__weak void hl_pci_iatu_init(void) {}
-__weak bool is_pcie_initialized(void) { return true; }
-__weak void hl_dev_pci_pll_wa(void) {}
-
-static void module_pci_do_setup(void)
-{
-	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
-		pci_dev_specific_setup();
-		pci_dev_flr_enable(true);
-	}
-
-	pci_sts.setup_done = true;
-}
-
-void module_pci_setup(void)
-{
-	if (!is_pcie_initialized())
-		return;
-
-	module_pci_do_setup();
-}
-
-#define DBG_TIMEOUT_US		1000000
-#define DBG_LINE_SZ		80
-
-static unsigned int dbg_cnt;
-static u64 dbg_ts;
-
-static void pci_init_wait_dbg(void)
-{
-	if (timeout_expired(dbg_ts, DBG_TIMEOUT_US)) {
-		preboot_puts_raw_dbg("*");
-		dbg_ts = timestamp_get_us();
-		dbg_cnt++;
-	}
-
-	if (dbg_cnt == DBG_LINE_SZ) {
-		preboot_puts_dbg("\n");
-		dbg_cnt = 0;
-	}
-}
-
-__weak u32 pci_dev_specific_hot_rst_check(void)
-{
-	return regmap_readl(pci.core_hot_rst_reg) & BIT(0);
-}
-
-__weak enum cmd_check_res dev_specific_module_pci_check(void)
-{
-	/*
-	 * don't read pci status regs until PRSTn is deasserted and pci blocks
-	 * are functional
-	 */
-	if (!is_pcie_initialized()) {
-		pci_init_wait_dbg();
-		return NOT_FOUND;
-	}
-
-	if (!pci_sts.setup_done)
-		module_pci_do_setup();
-
-	pci_sts.flr_reg_val = regmap_readl(pci.aux_flr_int_reg);
-	pci_sts.hot_rst_reg_val = pci_dev_specific_hot_rst_check();
-	pci_sts.rdlh_link_reg_val = regmap_readl(pci.aux_rdlh_link_up_reg);
-	pci_sts.smlh_link_reg_val = regmap_readl(pci.aux_smlh_link_up_reg);
-
-	if (!pci_sts.pci_pll_initialized)
-		return FOUND_SLOW_RETURN;
-
-	if (pci_sts.flr_reg_val == 1 || pci_sts.hot_rst_reg_val == 1)
-		return FOUND_FAST_RETURN;
-
-	if ((pci_sts.rdlh_link_reg_val && pci_sts.smlh_link_reg_val) !=
-	    pci_sts.link_is_up)
-		return FOUND_FAST_RETURN;
-
-	if (pci_sts.link_is_up)
-		pci_sts.mem_enabled = hl_pci_is_mem_enabled();
-
-	if (pci_sts.mem_enabled != pci_sts.mem_enabled_prev)
-		return FOUND_FAST_RETURN;
-
-	return NOT_FOUND;
-}
-
-enum cmd_check_res module_pci_check(void)
-{
-	if (hl_is_debug_mode(HL_DEBUG_HW_PCIE))
-		return NOT_FOUND;
-
-	return dev_specific_module_pci_check();
-}
-
-void module_pci_disable(void)
-{
-	if (!is_pcie_initialized())
-		return;
-
-	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
-		pci_dev_specific_disable();
-		pci_dev_flr_enable(false);
-	}
-
-	pci_sts.setup_done = false;
-}
-
-__weak void pci_dev_specific_run(void)
-{
-	unsigned int rst_src = hl_rst_src_get();
-
-	if (!pci_sts.pci_pll_initialized) {
-		if ((rst_src == HL_COLD_RST) || (rst_src == HL_PRSTN_RST))
-			hl_dev_pci_pll_wa();
-
-		pci_sts.pci_pll_initialized = true;
-	}
-}
-
-void module_pci_run(void)
-{
-	pci_dev_specific_run();
-}
-
-__weak void pci_dev_specific_post_link_up(void) {}
-
-void module_pci_run_fast(void)
-{
-	unsigned int rst_src;
-
-	if (pci_sts.flr_reg_val == 1) {
-		pci_dev_specific_flr();
-		preboot_puts("FLR received\n");
-	}
-
-	if (pci_sts.hot_rst_reg_val == 1) {
-		pci_dev_specific_hot_rst();
-		preboot_puts("HOT RST received\n");
-	}
-
-	if ((pci_sts.rdlh_link_reg_val && pci_sts.smlh_link_reg_val) !=
-	    pci_sts.link_is_up) {
-		pci_sts.link_is_up = !pci_sts.link_is_up;
-		preboot_puts_raw("PCIe link is ");
-		if (pci_sts.link_is_up) {
-			preboot_puts("up\n");
-			pci_dev_specific_post_link_up();
-		} else {
-			preboot_puts("down\n");
-			pci_sts.mem_enabled = false;
-		}
-	}
-
-	if (pci_sts.mem_enabled_prev != pci_sts.mem_enabled) {
-		pci_sts.mem_enabled_prev = pci_sts.mem_enabled;
-		rst_src = hl_rst_src_get();
-		if (pci_sts.mem_enabled && ((rst_src == HL_COLD_RST) ||
-					    (rst_src == HL_PRSTN_RST)))
-			hl_pci_iatu_init();
-
-		preboot_puts_raw("PCIe mem_en changed to: ");
-		preboot_puts_raw(pci_sts.mem_enabled ? "true" : "false");
-		preboot_puts("\n");
-	}
-}
-
-void module_pci_enable(void)
-{
-	if (!is_pcie_initialized())
-		return;
-
-	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
-		pci_dev_specific_enable();
-		pci_dev_flr_enable(true);
-	}
-
-	pci_sts.setup_done = true;
-}
-
-void module_pci_teardown(void)
-{
-	if (!is_pcie_initialized())
-		return;
-
-	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
-		pci_dev_specific_teardown();
-		pci_dev_flr_enable(false);
-	}
-
-	pci_sts.setup_done = false;
-}
-
-bool hl_pci_link_is_up(void)
-{
-	return pci_sts.link_is_up;
-}
